sympy/assumptions/cnf.py

```diff
@@ -0,0 +1,373 @@
+        return AND(*[to_NNF(x) for x in And.make_args(expr)])
+
+    if klass == 'Nand':
+        tmp = AND(*[to_NNF(x) for x in expr.args])
+        return ~tmp
+
+    if klass == 'Nor':
+        tmp = OR(*[to_NNF(x) for x in expr.args])
+        return ~tmp
+
+    if klass == 'Xor':
+        cnfs = []
+        for i in range(0, len(expr.args) + 1, 2):
+            for neg in combinations(expr.args, i):
+                clause = [~to_NNF(s) if s in neg else to_NNF(s)
+                          for s in expr.args]
+                cnfs.append(OR(*clause))
+        return AND(*cnfs)
+
+    if klass == 'Xnor':
+        cnfs = []
+        for i in range(0, len(expr.args) + 1, 2):
+            for neg in combinations(expr.args, i):
+                clause = [~to_NNF(s) if s in neg else to_NNF(s)
+                          for s in expr.args]
+                cnfs.append(OR(*clause))
+        return ~AND(*cnfs)
+
+    if klass == 'Implies':
+        L, R = map(to_NNF, expr.args)
+        return OR(~L, R)
+
+    if klass == 'Equivalent':
+        cnfs = []
+        for a, b in zip(expr.args, expr.args[1:]):
+            a = to_NNF(a)
+            b = to_NNF(b)
+            cnfs.append(OR(~a, b))
+        a = to_NNF(expr.args[-1])
+        b = to_NNF(expr.args[0])
+        cnfs.append(OR(~a, b))
+        return AND(*cnfs)
+
+    if klass == 'ITE':
+        L, M, R = map(CNF.to_CNF, expr.args)
+        return AND(OR(~L, M), OR(L, R))
+
+
+def distribute_AND_over_OR(expr):
+    """
+    Distributes AND over OR in the NNF expression.
+    Returns the result( Conjunctive Normal Form of expression)
+    as a CNF object.
+    """
+    if not isinstance(expr, (AND, OR)):
+        tmp = set()
+        tmp.add(frozenset((expr,)))
+        return CNF(tmp)
+
+    klass = type(expr).__name__
+    if klass == 'OR':
+        return CNF.all_or(*[distribute_AND_over_OR(arg)
+                            for arg in expr._args])
+
+    if klass == 'AND':
+        return CNF.all_and(*[distribute_AND_over_OR(arg)
+                             for arg in expr._args])
+
+
+class CNF(object):
+    """
+    Class to represent CNF of a Boolean expression.
+    Consists of set of clauses, which themselves are stored as
+    frozenset of Literal objects.
+    """
+    def __init__(self, clauses=None):
+        if not clauses:
+            clauses = set()
+        self.clauses = clauses
+
+    def add(self, prop):
+        clauses = CNF.to_CNF(prop).clauses
+        self.add_clauses(clauses)
+
+    def __str__(self):
+        s = ' & '.join(
+            ['(' + ' | '.join([str(lit) for lit in clause]) +')'
+            for clause in self.clauses]
+        )
+        return s
+
+    def extend(self, props):
+        for p in props:
+            self.add(p)
+        return self
+
+    def copy(self):
+        return CNF(set(self.clauses))
+
+    def add_clauses(self, clauses):
+        self.clauses |= clauses
+
+    @classmethod
+    def from_prop(cls, prop):
+        res = CNF()
```

Comment:
```suggestion
        res = cls()
```

Reply:
Does it have any benefit? AMAIK, this change may break things while inheritance. Though I don't think we may subclass CNF, I would better leave this if there is no benefit.

Reply:
The change is needed for it to be a proper class method so that it can be overridden by subclasses. It's best not to question things like that because there is no advantage to you in hard-coding the class and it has the potential to break things for someone else down the line.